### ISSUE TITLE
Gracefully shutting down a TLS server sometimes leads to the client not receive a response

### ISSUE THREAD

**Issue by chinedufn** [hyperium/hyper#3792]
```toml
hyper = { version = "=1.5.1", features = ["client", "http1"] }
hyper-util = { version = "=0.1.10", features = ["http1", "tokio", "server"] }
```

```text
MacBook Pro 16-inch, 2019
2.4 GHz 8-Core Intel Core i9
```

I'm observing errors while testing graceful shutdown of a hyper server.

Sometimes, after gracefully shutting down a TLS connection the client will never receive a response.
This only happens for TLS connections. The graceful shutdown process is always successful for non-TLS connections.

Given the following testing steps:
- Start the server.
- Send many concurrent requests to the server
- Wait for any of the requests to receive a response.
    (Since the request handler has a random sleep between 5-10ms we can be reasonably confident
    that when we receive a response there are still some other requests that are in-progress.)
- Tell the server to shut down. We expect that there are still some in-progress requests.
- Assert that we receive a 200 OK response for each request.
- This means that the graceful shutdown process was successful

When the hyper server is not using TLS, the test pass.
When the hyper server is using TLS, the test fails.

I've created a repository that reproduces the issue https://github.com/chinedufn/hyper-tls-graceful-shutdown-issue
```sh
git clone https://github.com/chinedufn/hyper-tls-graceful-shutdown-issue
cd hyper-tls-graceful-shutdown-issue

# This always passes
cargo test -- graceful_shutdown_without_tls

# This always fails
cargo test -- graceful_shutdown_with_tls
```

**Comment by seanmonstar** [comment]
Have you been able to trace what is happening? One thing that sort of sounds like is that the TLS stream perhaps hasn't flushed the response before closing?

**Comment by seanmonstar** [comment]
But do you know for sure all the requests have indeed started? Or could the shutdown be triggered just before hyper has been able to see the request bytes?

**Comment by chinedufn** [comment]
> Or could the shutdown be triggered just before hyper has been able to see the request bytes?

Seems to be the problem. Most of this comment is how I arrived at that. Skip to the end for a potential solution.

---

Did a bit of `hyper=trace,hyper_util=trace` tracing before opening the issue but nothing jumped out at me.

Hmm, so if I sleep for `1ms` before dropping the `watch::Receiver` that serves as the shut down signal then the test passes.
As in, a sleep before this line: https://github.com/chinedufn/hyper-tls-graceful-shutdown-issue/blob/7c3d52f54e839096022a3e9b7b478ad0a635293a/src/lib.rs#L42

Inlining the snippet code here for convenience:
```rust
    loop {
        tokio::select! {
            _ = &mut shutdown_receiver => {
                // ADDING THIS SLEEP MAKES THE TEST PASS
                tokio::time::sleep(std::time::Duration::from_millis(1)).await;
                drop(shut_down_connections_rx);
                break;
            }
            conn = tcp_listener.accept() => {
                tokio::spawn(
                    handle_tcp_conn(
                        conn,
                        wait_for_request_to_complete_rx.clone(),
                        shut_down_connections_tx.clone(),
                        tls_config
                    )
                );
            }
        }
    }
```

The `drop(shut_down_connections_rx)` shut down signal causes this `Connection::gracful_shutdown` `tokio::select!` branch to be selected.
https://github.com/chinedufn/hyper-tls-graceful-shutdown-issue/blob/7c3d52f54e839096022a3e9b7b478ad0a635293a/src/lib.rs#L145-L159

Inlining the snippet here for convenience:
```rust
    tokio::select! {
        result = conn.as_mut() => {
            if let Err(err) = result {
                dbg!(err);
            }
        }
        _ = should_shut_down_connection => {
            conn.as_mut().graceful_shutdown();
            let result = conn.as_mut().await;
            if let Err(err) = result {
                dbg!(err);
            }
        }
    };
```

## Key Points

The test passes if we sleep for a millisecond before sending on the channel that leads to `conn.as_mut().graceful_shutdown();` getting called for all open connections.
i.e. if we sleep for one millisecond right before this line: https://github.com/chinedufn/hyper-tls-graceful-shutdown-issue/blob/7c3d52f54e839096022a3e9b7b478ad0a635293a/src/lib.rs#L42

If I instead move the sleep to just before the `conn.as_mut().graceful_shutdown();` line, the test fails, even if we sleep for 1, 5 seconds, 50milliseconds or seemingly any other amount of time.
( i.e. if we sleep for five seconds right before this line the test will fail -> https://github.com/chinedufn/hyper-tls-graceful-shutdown-issue/blob/7c3d52f54e839096022a3e9b7b478ad0a635293a/src/lib.rs#L152 )
( I also confirmed that sleeping for 50 milliseconds leads the test to fail.)

This suggests that the problem occurs when we call `Connection::graceful_shutdown` before we've started polling the connection and receiving bytes.

It looks like `graceful_shutdown` calls `disable_keepalive`, and `disable_keepalive` closes the connection if no bytes have been received.
https://github.com/hyperium/hyper/blob/master/src/proto/h1/dispatch.rs#L90-L100

> Or could the shutdown be triggered just before hyper has been able to see the request bytes?

Yeah seems like this is the problem.

## Problem
Currently, if a user opens a TCP connection to a server and the server calls `Connection::graceful_shutdown` before any bytes have been received, the TCP connection will be closed.

This means that if the client has just begun transmitting packets, but the server has not received them, the client will get an error.
This is not a graceful shutdown since the client was not made aware that the connection was going to be closed.


## Potential Solutions

One way to avoid such an error would be to something like:

1. Server somehow indicates to client that the connection will close
2. Client receives indication that server is planning to close the connection
3. If the client has not yet sent any data, server waits for `Duration` "W" (configurable) to see if the client sends any packets
4. If the client DOES NOT send any packets after the wait `Duration` "W", close the connection
5. If the client DOES send packets before the wait `Duration` "W", receive the bytes and then close the connection

I'm unfamiliar with the TCP spec, but from some quick searching it the `FIN` segment might be useful here?
I can do more research if the above steps seem like a good path forward.

Is this a viable path forward?

**Comment by chinedufn** [comment]
Are you aware that an HTTP keep alive is different from a TCP keep alive? https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Keep-Alive

If the server disables HTTP Keep-Alive it is indicating to the client that at most one more HTTP request will be processed.

I ask because I'm not seeing how the steady stream of packets concept is relevant here.
If a single request takes more than some amount of time the hyper user can just close the connection and shut down their server.
And disabling HTTP keep-alive means that no new requests will be accepted.
So at the time of shut down we only care about whether to finish handling the current/next request.

---

The idea is that if a connection has been opened and 0 requests have been received by the server, either:

- the client is well-behaved and is about to send the request bytes to the server
- OR the client is well-behaved but their connection closed and the bytes will never reach the server
- OR the client is an attacker and wants to hold the connection open indefinitely

I'm suggesting that it should be possible to accept one more request from a well-behaved client whose bytes we are, say, milliseconds away from beginning to process.

Whereas right now if we would have received final request's the bytes in 5 milliseconds `hyper` doesn't care and will close the connection. (or at least that's my current understanding. not familiar enough with TCP / HTTP to be sure that that's the case here).

**Comment by SFBdragon** [comment]
> Are you aware that an HTTP keep alive is different from a TCP keep alive?

Not familiar with the details, just read up on it, doesn't seem like anything about the argument I put forward should change in light of this (maybe I'm missing something!). The argument was intended to be robust regardless of any other details, and I think it still is.

If this keepalive stuff implies that the server is actually doing something against the standards by sending a `RST` without warning, then ignore everything I've said. Let's keep `hyper` conformant. But I'm under the impression that `hyper`'s current behavior is fine, at least on the TCP layer. If there's something wrong on the HTTP layer, I'm missing it.

> I ask because I'm not seeing how the steady stream of packets concept is relevant here.

I'll re-word my argument above. It's a bit more sloppy, but a bit more concrete:

- Let's say we get a packet every 10ms, indefinitely.
- Let's say graceful shutdown is called.
- Should we wait for 10ms to handle the next packet? I claim there's no point in any practical situation. If we think it's important to handle the next packet, why? Let's suppose there's a good reason. We handle it. Now there's another packet in 10ms. Surely the reason from before applies again. So we handle the next packet. etc. etc. etc. My point is that there's no good reason to wait 10ms.

Similarly, again making the steady-stream assumption:
- Server receives a connection. In 10ms the server received a request from the connection, for each connection. A connection comes in every 10ms. Each connection is closed soon afterward.
- Should we wait 10ms to handle the next request on the already-alive connection? No. If there's any good argument to do so, then we handle it, but then the next connection opens before the previous one is done. And the same ad-infinitum reasoning goes.

If this reasoning seems rather abstract, that's fair and I wish I could do better (in a pragmatic sense), but the argument nonetheless implies that there's no real case where this should matter, and this is just a very abstract way of making that point (recursive induction, then a reducto-ad-absurdum). [Then why do it this way? So that it's robust.]

If you've got a real use case in mind where this change in behavior makes a crucial difference, then I must have made a mistake somewhere.

---

> I'm suggesting that it should be possible to accept one more request from a well-behaved client whose bytes we are, say, milliseconds away from beginning to process.

I agree, this will usually be the case. I don't think it would be bad to do this, assuming the cost is zero (increased complexity in the `hyper` API, for example). But it won't help with any real use case.

>  This issue is about figuring out the best way to get those bytes (be it via hyper or userland)

This is assuming that there's a good reason to get those bytes. I've made the case that there isn't one.

> If that holds it's reasonable for a maintainer of a hyper based server to want to disable HTTP Keep-Alive and then wait a couple seconds before terminating the connection.

Only if it makes a difference to clients or the server in general. But it does not by the above.

---

What I fear is that we're inventing a use-case that doesn't exist outside of our one test, and expecting `hyper` to support that.

**Comment by chinedufn** [comment]
> Should we wait for 10ms to handle the next packet?

This line of reasoning misunderstands what an HTTP graceful shutdown is.

An HTTP graceful shutdown is basically "if there is a request that is already in progress, finish responding and then close the connection".

So, yes you should continue receiving those packets every 10ms until so timeout is reached. In a production system that timeout might be in the neighborhood of 30 seconds.

> Should we wait for 10ms to handle the next packet?

Today, if `Connection::graceful_shutdown` is callef after `hyper` has already begun receiving packets it will continue to until the request is complete. It will close the connection after the response has been sent.

https://github.com/hyperium/hyper/blob/4ffaad53c78572c500584e0cb5d76ae6ffc6adb6/src/proto/h1/dispatch.rs#L90-L100

So, `hyper` already exhibits this "continue reading packets" behavior under some circumstances.

> A connection is established every 10ms. 

This does not apply. During a graceful shutdown you would stop accepting new TCP connections.

Here is an example from this issue's linked repository: https://github.com/chinedufn/hyper-tls-graceful-shutdown-issue/blob/7c3d52f54e839096022a3e9b7b478ad0a635293a/src/lib.rs#L39-L56

> If you've got a real use case in mind where this change in behavior makes a crucial difference, then I must have made a mistake somewhere.

It's hard for me to try to provide new use cases when your argument is already arguing against how `hyper` currently works. That leaves me thinking that there's a misunderstanding.

`hyper` already does everything that you're claiming is bad (continuing to accept packets after `Connection::graceful_shutdown` is called).

It does this because that is was a graceful shutdown is. Continuing to process a request before closing the connection.

**Comment by SFBdragon** [comment]
I'm coming back to this having caught up some sleep and with some further discussion. My main concerns were that the benefit was negligible and the only cases it mattered were artificial. I disagree with these two conclusions now.

Here's my current take of why I think this should be fixed:
- Because this only happens with TLS configured, this appears to be an inconsistency that's probably better off being fixed. (Better documentation about `graceful_shutdown`'s intended behavior in general would be valuable, even if it's noncommital about what may change in the future, though that would be useful too.)
- Assuming the server is shut down and clients will fail to make requests from there, onwards is an assumption I implicitly made that doesn't necessarily hold. For example, shutting down a main server in favor of a backup, or vice-versa. In which case, changing the `graceful_shutdown` as pushed for by @chinedufn would help avoid any error conditions as far as the client is concerned. This will likely be useful to us in the future.
- Furthermore, while in other cases, being aggressive about shutting down connections without a request received yet isn't better in a make-or-break sense than being more lenient on recently opened connections, it would be marginally better.
- I was also assuming that we always controlled the client. This is not the case, and it would be better to reduce the potential error conditions we could be exposing clients to.

Therefore, I think this issue is worth solving. Sorry about all the noise I've added to the issue, feel free to hide my comments to keep things organized.

**Comment by seanmonstar** [comment]
I do agree with some of the points brought up:

1. A client does need to handle that the connection could die, or that the server could close it on a timeout.
2. If we add a timeout, it only helps in _this_ specific case. No matter how long the timeout is, there will always be a case where if we had waited just a little longer, we'd be able to read the request.

It seems likely that in this case, the time it takes to write/encrypt/read/decrypt is just enough that the tiny timeout used can beat it.

**Comment by chinedufn** [comment]
Definitely agree that a timeout wouldn't solve 100% of cases.

My thinking is that if a client opens a connection they're either:

1. An attacker that would love to hold the connection open forever
2. A non-malicious user whose bytes we're about to receive in some small amount of time
3. A non-malicious user whose bytes might take a long time to reach us, or might never reach us 

I'm not concerned about making things easier for the attacker (1).

I'd love to make things (slightly) easier for the non-malicious user who was about to give me bytes (2). I can save them a retry by handling the request.

But, yes, for case (3) if it takes to long to see your bytes we'll need to close the connection. Can't wait forever.

So, case (2) is what I'm interested in handling.

I am operating under the assumption (no data on this currently) that with a timeout of, say, 1-2 seconds, I could handle the overwhelming majority of "I was about to give you bytes but you closed the connection on me!" scenarios.

This means that my users see fewer errors, at nearly no expense to my server (I'm already comfortable taking up to 30 seconds to handle any in-progress requests before closing all connections. This 1-2 seconds would just come from that 30 second budget.)

Only trade-off I can imagine is `hyper` becoming more complex in the name of solving an edge case that a client should probably already be prepared for.

I'd say that saving the extra retry request for the client is worth it if the cost to `hyper`'s API surface area is negligible (or maybe it's already possible/easy for a user to disable keep alive on the HTTP connection?).

The use case that I imagine is a server that sees heavy consistent traffic and is gracefully shut down often (say, every hour when the maintainers redeploy a new version of it).
Solving this means the server's client's will see fewer errors than they otherwise would have.

---

Are you open to solving this "problem" (I recognize that one might argue that it isn't a problem and the client should just be expected to retry) in `hyper`, assuming there was a good way to do so?

Do any of the potential solutions that I've left in my previous comment seem reasonable?

I'm happy to do the work required to land a solution to the "decrease the number of connections that get closed when we could have reasonably waited for a bit to get the data" problem (to repeat, yup I know we can't wait forever. I just want to reduce errors that users experience and recognize that I cannot fully eliminate them).

**Comment by chinedufn** [comment]
Alright I took a look at the `hyper` source and I think I have a reasonable solution.

To recap, the main problem is that `hyper`'s graceful shutdown implementation currently leads to more errors than necessary.

This means that changing the graceful shutdown implementation will lead to a better experience for end clients, meaning that developers using hyper to write web servers will be able to build (slightly) more reliable systems.

---

Some background / related issues:

In Jan 1 2022 an issue called `Connection::graceful_shutdown always waits for the first request` was opened. https://github.com/hyperium/hyper/issues/2730

At the time, if you called `Connection::graceful_shutdown`, but the client never transmitted any bytes, the connection would remain open forever.

In July 2023 `fix(http1): http1 server graceful shutdown fix` https://github.com/hyperium/hyper/pull/3261 addressed that issue by making `Connection::graceful_shutdown` immediately close the connection if no bytes have been received.

---

Currently, calling `Connection::graceful_shutdown` will cause some clients to receive an error despite the server being perfectly willing and able to handle the request.

hyper should give server authors control over the grace period for graceful shutdowns so that they can decide what the best experience is for their end clients.

Similar to how hyper currently allows the server author to control the header read timeout using `set_http1_header_read_timeout`.

The same level of control should exist for graceful shutdowns.

---

## Proposed Changes to Hyper

Here is how we can solve this graceful shutdown problem (clients unnecessarily getting errors when waiting a second or two before terminating an unwritten connection would have given nearly every connection a chance of being gracefully handled).

Currently, `Connection::graceful_shutdown` will immediately close the connection if no bytes have been written.

Solution -> `Connection::graceful_shutdown` will immediately disable keepalive, but then schedule a future for closing the connection, similar to how the `h1_header_read_timeout_fut` works: https://github.com/hyperium/hyper/blob/c86a6bcb4acb0f92e731ea2e4c1e4a839248a600/src/proto/h1/conn.rs#L238-L272

So, something like a `set_http1_graceful_shutdown_read_timeout` (that's a made up messy name).

By default the `http1_graceful_shutdown_read_timeout` duration is `0` seconds. This preserves the current behavior, meaning that `hyper` users experience zero changes in behavior.

Hyper users can use `set_http1_graceful_shutdown_read_timeout` to increase that timeout (i.e. to 2 seconds, if they like), which should let them reduce the number of errors that they experience.

This allows us to make this change now since we can fully preserve the current behavior entirely.

Then we can take as much time as we like to decide whether the default should be changed from `0` seconds to something else. (I haven't thought much about this ... but 0 seems wrong since it guarantees that a high traffic server's clients will get errors whenever the server is taken down ... something like 1-5 seconds would eliminate nearly all of those errors... so that's probably a better default..)

To summarize the benefits of this solution:

- non-breaking API change

- preserves the current behavior (albeit flawed behavior since I'd expect most people running production web servers to prefer a default of, say, 2 seconds over 0 seconds)

## Next Steps

I'm planning to work on implemeneting this this week. Planning to use some of the tests in `tests/server.rs` as a guide for how to introduce new tests. i.e. https://github.com/hyperium/hyper/blob/c86a6bcb4acb0f92e731ea2e4c1e4a839248a600/tests/server.rs#L1542-L1561

Please let me know if you see any issues with this proposed solution.

**Comment by chinedufn** [comment]
Looks like the HTTP/2 spec has support for preventing the problem that this issue is about.

>  A client that is unable to retry requests loses all requests that are
>  in flight when the server closes the connection.  This is especially
>  true for intermediaries that might not be serving clients using
>  HTTP/2.  A server that is attempting to gracefully shut down a
>  connection SHOULD send an initial GOAWAY frame with the last stream
>  identifier set to 2^31-1 and a NO_ERROR code.  This signals to the
>  client that a shutdown is imminent and that initiating further
>  requests is prohibited.  After allowing time for any in-flight stream
>  creation (at least one round-trip time), the server can send another
>  GOAWAY frame with an updated last stream identifier.  This ensures
>  that a connection can be cleanly shut down without losing requests.

via: https://datatracker.ietf.org/doc/html/rfc7540#section-6.8

`hyper'`s HTTP/2 graceful shutdowns are implemented properly.

So, looks like only HTTP/1 connections are experience this unnecessarily-dropping-requests-from-well-behaved-clients issue.

https://github.com/hyperium/h2/blob/3bce93e9b0dad742ffbf62d25f9607ef7651a70c/src/proto/connection.rs#L582-L605

`hyper`'s HTTP/2 implementation follows the spec's recommendations of waiting for at least one round-trip time before ceasing to accept new request streams.

---

Note that since HTTP/1's disable keepalive means that only one more request will be processed, we can't mimic this HTTP/2 behavior of allowing many not-yet-received requests
to get processed.

As in, if the client sent 10 requests that did not yet reach the server when `.graceful_shutdown()` was called, in HTTP/2 we can still process all ten, whereas in HTTP/1 we'll only
process whichever one we receive first.

---

Current progress: I wrote some tests and got them passing. Still need to clean up the code. Planning to submit a PR this week.

The change to the public API is a `Connection::graceful_shutdown_with_config` method that lets you configure a `Duration` for HTTP/1 connections to wait before closing the inactive connection.

The existing `Connection::graceful_shutdown` method sets the wait period to 0 seconds.
No timer is required when the Duration is 0 (making this a fully non-breaking change).

**Comment by chinedufn** [comment]
I've confirmed that HTTP/2 hyper servers do not suffer from this issue.

I confirmed this by enabling the `rust-tls` feature on `reqwest` and then using `ClientBuilder::use_rustls_tls` to make `reqwest` use HTTP/2.

When I use `use_rustls_tls` this issue cannot be reproduced. When I comment out `use_rustls_tls` this issue happens again.
